SourceX/stash.cpp

```diff
@@ -0,0 +1,447 @@
+		for (int i = 0; i < NUM_STASH_GRID_ELEM; i++) {
+			StashList[i]._itype = ITYPE_NONE;
+		}
+		_pNumStash = 0;
+	}
+}
+
+InvXY getStashGridXY(int i)
+{
+	return {
+		stashGridStart.X + (i % stashCols) * stashGridOffsets.X, stashGridStart.Y + (i / stashRows) * stashGridOffsets.Y
+	};
+}
+
+std::string GetStashPath()
+{
+	return GetPrefPath() + "stash.lol";
+}
+
+int GetStashSlotFromMouse(int mx, int my)
+{
+	for (int r = 0; (DWORD)r < NUM_STASH_GRID_ELEM; r++) {
+		// check which inventory rectangle the mouse is in, if any
+		InvXY invrect = getStashGridXY(r);
+		if (mx >= invrect.X && mx < invrect.X + (INV_SLOT_SIZE_PX + 1) && my >= invrect.Y - (INV_SLOT_SIZE_PX + 1) && my < invrect.Y) {
+			return r;
+		}
+	}
+	return -1;
+}
+
+char CheckStashHLight()
+{
+	if (!stash)
+		return -1;
+	int r, ii, nGold;
+	ItemStruct *pi;
+	char rv;
+
+	r = GetStashSlotFromMouse(MouseX, MouseY);
+
+	if (r == -1)
+		return -1;
+
+	rv = -1;
+	infoclr = COL_WHITE;
+	pi = NULL;
+	ClearPanel();
+	r = abs(StashGrid[r]);
+	if (r == 0)
+		return -1;
+	ii = r - 1;
+	rv = ii;
+	pi = &StashList[ii];
+
+	if (pi->_itype == ITYPE_NONE)
+		return -1;
+
+	if (pi->_itype == ITYPE_GOLD) {
+		nGold = pi->_ivalue;
+		sprintf(infostr, "%i gold %s", nGold, get_pieces_str(nGold));
+	} else {
+		if (pi->_iMagical == ITEM_QUALITY_MAGIC) {
+			infoclr = COL_BLUE;
+		} else if (pi->_iMagical == ITEM_QUALITY_UNIQUE) {
+			infoclr = COL_GOLD;
+		}
+		strcpy(infostr, pi->_iName);
+		if (pi->_iIdentified) {
+			strcpy(infostr, pi->_iIName);
+			PrintItemDetails(pi);
+		} else {
+			PrintItemDur(pi);
+		}
+	}
+
+	return rv;
+}
+
+bool SaveStash()
+{
+	FILE *pFile = fopen(GetStashPath().c_str(), "wb");
+
+	if (pFile == NULL)
+		return false;
+
+	BYTE *tbuffold = tbuff;
+	DWORD dwLen = (sizeof(ItemStruct) * NUM_STASH_GRID_ELEM) + sizeof(char) * NUM_STASH_GRID_ELEM + sizeof(int);
+	BYTE *tmp = (BYTE *) malloc(dwLen);
+	memset(tmp, 0, dwLen);
+	tbuff = tmp;
+	CopyInt(&_pNumStash, tbuff);
+	SaveItems(StashList, NUM_STASH_GRID_ELEM);
+	CopyBytes(StashGrid, NUM_STASH_GRID_ELEM, tbuff);
+	fwrite(tmp, sizeof(char), dwLen, pFile);
+	fclose(pFile);
+	free(tmp);
+	tbuff = tbuffold;
+	return true;
+}
+
+bool LoadStash()
+{
+	FILE *pFile;
+	unsigned long lSize;
+	size_t result;
+
+	pFile = fopen(GetStashPath().c_str(), "rb");
+	if (pFile == NULL)
+		return false;
+
+	// obtain file size:
+	fseek(pFile, 0, SEEK_END);
+	lSize = ftell(pFile);
+	rewind(pFile);
+
+	// allocate memory to contain the whole file:
+	BYTE *tbuffold = tbuff;
+	BYTE *tmp = (BYTE *)malloc(lSize);
+	if (tmp == NULL) {
+		fclose(pFile);
+		return false;
+	}
+	memset(tmp, 0, lSize);
+
+	// copy the file into the buffer:
+	result = fread(tmp, 1, lSize, pFile);
+	if (result != lSize) {
+		free(tmp);
+		fclose(pFile);
+		return false;
+	}
+	/* the whole file is now loaded in the memory buffer. */
+	// terminate
+	fclose(pFile);
+
+	tbuff = tmp;
+	CopyInt(tbuff, &_pNumStash);
+	LoadItems(NUM_STASH_GRID_ELEM, StashList);

```

Comment:
Don't forget to set gbIsHellfireSavegame before calling the savegame functions.

Reply:
isn't it set already?

Reply:
It gets set to different values depending on what operation is about to be done. So you need to set it to the appropriate value before calling any of the loading routines to be safe.

reading hellfire save = true
reading diablo save = false
networking diablo mode = false
networking hellfire mode = true
cornerstone = true

```suggestion
	gbIsHellfireSaveGame = gbIsHellfire;
	LoadItems(NUM_STASH_GRID_ELEM, StashList);
```